src/FreeAge/common/type_info.cpp

```diff
@@ -0,0 +1,621 @@
+// Copyright 2020 The FreeAge authors
+// This file is part of FreeAge, licensed under the new BSD license.
+// See the COPYING file in the project root for the license text.
+
+#include "FreeAge/common/free_age.hpp"
+#include "FreeAge/common/unit_types.hpp"
+#include "FreeAge/common/building_types.hpp"
+#include "FreeAge/common/player.hpp"
+#include "FreeAge/common/damage.hpp"
+#include "FreeAge/common/resources.hpp"
+#include "FreeAge/common/object_types.hpp"
+
+// Naming conversions:
+// *Stats: collection of values
+// *Info: the base *Stats, a list of modications on it and the resulting *Stats
+
+// Player has a CivilizationInfo, a vector<UnitTypeInfo>, a vector<BuildingTypeInfo> and a vector<TechnologyInfo>
+// The game data provide the base UnitTypeStats, BuildingTypeStats, TechnologyStats and CivilizationStats
```

Comment:
Even if it's not possible/worth it loading it from the original game assets, I think it would be good to store this in some sort of data file. It could also be advantageous from a community perspective, as people who can't code could contribute to the project by filling these files.

Reply:
Technologies I think are modeled with a huge switch, no data there... 

I believe loading all everything from original would need a lot of work. I think the best would be to (optionally) have a tool that could extract the information only about the basic unit/building stats into the FreeAge data format the first time, however, like @grst said, is it worth it?

I suggest to have a **data.cpp** file which creates all the data using the *Stats structs (and whatever else is added for dependencies and building interfaces). In the future, if multiple game data are 100% needed and need to be stored outside of the code, a simple binary dump of the data could be implemented (too much work for now without any visible effect)

Bonus: We could add a bunch of good old c macros in the **data.cpp** and pretend is a FreeAge's text based data format (with type checking)

Reply:
Why not just go for some standard format, such as JSON or [TOML](https://github.com/toml-lang/toml)?

Reply:
Imagine this in json:
```
  // Team bonus: Barracks work 20% faster.
  age[3].teamBonus.addModification(
    FilterOfBuildingsByType(BuildingType::Barracks),
    Modification(Type::WorkRate, Operation::MultAdd, 20)
  )
```
It would just be ugly ** + you wouldn't have an ide helping you out (autocomplete, immediate feedback eg. "Type::Work does not exist") + you would need to parse all the ids (use strings: ok, but I dont like it**) + to compact the data you need something like functions (could use json objects with a type and a list of arguments: ok, but I dont like it**)

** I and Json have been fighting for a lot of years now, I don't want to talk about it, am not subjective, **just ignore this response**

Reply:
fair enough

Reply:
My impression is that the drawbacks of using JSON would not be severe, but I also don't have much experience using it :) I guess that it would still be easy to add a custom data format later if we start with definitions in code now. Whether compatibility to the original game files should be aimed at seems like it might be harder to change later (as this might imply replicating the game data structures in the same way). As far as I can see, there might be some consensus that this is not necessarily worth it.